rfcs/0074-tvmscript-unified-printer.md

```diff
@@ -0,0 +1,487 @@
+- Feature Name: TUNIP: TVMScript Unified Printer
+- Start Date: 05/25/2022
+- RFC PR: [apache/tvm-rfcs#0000](https://github.com/apache/tvm-rfcs/pull/0000)
+- GitHub Issue: [apache/tvm#0000](https://github.com/apache/tvm/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes to modularize and infrastructuralize the existing TVMScript
+printer, to develop unified printing mechanism across TVM stack, where TIR,
+Relax and any future vendor-specific IR are all treated equally as dialects and
+could be printed together.
+
+# Motivation
+[motivation]: #motivation
+
+TVMScript, as a roundtrippable python-based text format, is
+the central piece of TVM performance productivity. As the frontend of TVM, it
+enables end users to directly construct the TVM IR, either TIR or Relax, in a
+pragmatic approach. From Relax to MetaSchedule and TIR, TVMScript enables
+inspectability and reproducibility at any level of compilation and
+optimization. Furthermore, based on TVMScript, developers are empowered to
+intercept, manipulate and customize the compiler behavior in a principled way.
+
+While TVMScript is gaining traction and buy-in from the open source community,
+the TVMScript printer suffers from multiple profound design issues:
+Not supporting IR fragment printing requires users to jump in-between
+TVMScript syntax and TIRText syntax; The lack of modularity leads to practical
+inability to scale up to and maintain multiple IRs without engineering
+conflicts; Enhancing co-existence of multi-level IRs often leads to
+re-engineering of existing features.
+
+**Goal.** This RFC introduces Tvmscript UNIfied Printer (TUNIP), a systematic
+redesign to
+address those engineering, usability and scalability issues above. The goal of
+this re-design includes:
+
+**Goal 1 [Unified Representation].** Become the unified roundtrippable
+representation of TIR and Relax, allowing systematic mixing of IRs or IR
+fragments (Relax + TIR) in the same IRModule in the target language (for
+example, python, C++). 
+
+The current TVMScript printer only supports TIR. An experimental printer is
+being developed at
+[https://github.com/tlc-pack/relax/blob/relax/src/printer/relax_script_printer.cc](https://github.com/tlc-pack/relax/blob/relax/src/printer/relax_script_printer.cc)
+for Relax, which needs to handle cases where TIR is embedded in Relax (for
+instance, https://github.com/tlc-pack/relax/pull/149). By having a unified
+approach for printing IR tree to TVMScript, we will be able to implement a
+fully-fledged printer for Relax and TIR with minimal effort.
```

Comment:
This is a bit vague. Let's make the following points here:
- Currently TVMScript is designed specifically for printing TIR, and printing multiple dialects together was not a design goal at that time;
- Therefore, supporting Relax requires ad-hoc hacks around the system
- The RFC address this issue by blah

Reply:
Done